src/main/java/org/radarcns/data/FileCache.java

```diff
@@ -42,21 +58,35 @@
      * @param path path to cache.
      * @param record example record to create converter from, this is not written to path.
      * @param gzip whether to gzip the records
-     * @throws IOException
+     * @throws IOException if the file and/or temporary files cannot be correctly read or written to.
      */
     public FileCache(RecordConverterFactory converterFactory, Path path,
-            GenericRecord record, boolean gzip) throws IOException {
+            GenericRecord record, boolean gzip, Path tmpDir) throws IOException {
         this.path = path;
         boolean fileIsNew = !Files.exists(path) || Files.size(path) == 0;
+        OutputStream outFile;
+        if (tmpDir == null) {
+            this.tmpPath = null;
+            outFile = Files.newOutputStream(path, StandardOpenOption.APPEND, StandardOpenOption.CREATE);
+        } else {
+            this.tmpPath = Files.createTempFile(tmpDir, path.getFileName().toString(),
+                    gzip ? ".tmp.gz" : ".tmp");
+            outFile = Files.newOutputStream(tmpPath, StandardOpenOption.WRITE);
```

Comment:
What happens here if the tmpfile already exists and is corrupt ? 
I think using the default with no options will be better - 
``` 
javadoc -
If no options are present then this method works as if the CREATE, TRUNCATE_EXISTING, and WRITE options are present. In other words, it opens the file for writing, creating the file if it doesn't exist, or initially truncating an existing regular-file to a size of 0 if it exists.
```

Reply:
Good point. I just realised that in addition, I'll need to code some logic for when the input file is corrupt.

Reply:
Yeah you can catch ZipException here and delete that file i guess. 
I have seen only the following exception when it tries to append to corrupt files --
```
java.util.zip.ZipException: invalid block type
	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:164)
	at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:117)
	at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:122)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at org.radarcns.util.RecordConverterFactory.testSortedAndUnique(RecordConverterFactory.java:166)
	at org.radarcns.util.RecordConverterFactory.sortUnique(RecordConverterFactory.java:73)
	at org.radarcns.util.FileCacheStore.ensureCapacity(FileCacheStore.java:108)
	at org.radarcns.util.FileCacheStore.writeRecord(FileCacheStore.java:78)
	at org.radarcns.RestructureAvroRecords.writeRecord(RestructureAvroRecords.java:287)
	at org.radarcns.RestructureAvroRecords.processFile(RestructureAvroRecords.java:243)
	at org.radarcns.RestructureAvroRecords.start(RestructureAvroRecords.java:186)
	at org.radarcns.RestructureAvroRecords.main(RestructureAvroRecords.java:93)
```